### ISSUE TITLE
SHACL data-type constraints not enforced when using http-api-gateway

### ISSUE THREAD

**Issue by aaj3f** [fluree/http-api-gateway#36]
## Description

My guess is this is true for SHACL enforcement more broadly, but only started testing with the data type constraint first. A basic SHACL constraint requiring a property on a particular class to be a `string` will succeed when using a REPL against db (i.e. they throw a data type error when trying to transact an integer against that property), but when using `http-api-gateway` the same invalid transaction is accepted and committed. 

The following is the set of invocations issued against the REPL:
```clojure
  (def conn         @(fluree/connect {:method       :file
                                      :storage-path "data"
                                      :defaults     {:context {:id     "@id"
                                                               :type   "@type"
                                                               :xsd    "http://www.w3.org/2001/XMLSchema#"
                                                               :rdf    "http://www.w3.org/1999/02/22-rdf-syntax-ns#"
                                                               :rdfs   "http://www.w3.org/2000/01/rdf-schema#"
                                                               :sh     "http://www.w3.org/ns/shacl#"
                                                               :schema "http://schema.org/"
                                                               :skos   "http://www.w3.org/2008/05/skos#"
                                                               :wiki   "https://www.wikidata.org/wiki/"
                                                               :f      "https://ns.flur.ee/ledger#"
                                                               :ex     "http://example.org/ns/"}}}))

  (def ledger       @(fluree/create conn "shacl/data-type"))

  (def ledger @(fluree/load conn "shacl/data-type"))

  (def db           @(fluree/stage
                      (fluree/db ledger)
                      {:id             :ex/UserShape,
                       :type           [:sh/NodeShape],
                       :sh/targetClass :ex/User
                       :sh/property    [{:sh/path     :schema/name
                                         :sh/datatype :xsd/string}]}))
  
  @(fluree/commit! ledger db)
```

The following is /create message sent to `http-api-gateway`:
```json
{
    "ledger": "shacl/data-type",
    "context": {
        "id": "@id",
        "type": "@type",
        "xsd": "http://www.w3.org/2001/XMLSchema#",
        "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
        "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
        "sh": "http://www.w3.org/ns/shacl#",
        "schema": "http://schema.org/",
        "skos": "http://www.w3.org/2008/05/skos#",
        "wiki": "https://www.wikidata.org/wiki/",
        "f": "https://ns.flur.ee/ledger#",
        "ex": "http://example.org/ns/"
    },
    "txn": {
        "@id": "ex:UserShape",
        "@type": ["sh:NodeShape"],
        "sh:targetClass": "ex:User",
        "sh:property": [
            {
                "sh:path": "schema:name",
                "sh:datatype": "xsd:string"
            }
        ]
    }
}
```

Interestingly, the two underlying commit files persisted by each are noticeably different
db-commit: https://gist.github.com/aaj3f/af3cdbcc11de97c4dbec23b99ebff608
http-commit: https://gist.github.com/aaj3f/bce1f66cc18388775efa949001907231

Notice in particular that only db-commit has the following properties defined as pointing to nodes/`@id` values:
```
    ...,
    "sh:datatype": {
      "@type": "@id"
    },
    "sh:path": {
      "@type": "@id"
    },
    ...
    "sh:targetClass": {
      "@type": "@id"
    },
```

**Comment by cap10morgan** [comment]
One thing to note that @mpoffald helped me realize (unless I'm mistaken haha): the combination of `"type": "@type"` in the context and having `"@type"` in the txn is invalid. Unlike the other context assignments (which add a shortcut you can use or ignore at your discretion AFAIUI), the key assigned to `"@type"` is special in that it replaces the `type-key` value in the context. If you assign `"type"` to `"@type"` in your context then you _have_ to use `"type"` in your transactions.

**Comment by mpoffald** [comment]
Yeah, that is causing us to not detect the type when we `expand`. Is that illegal JSON-LD, or is that just a limitation of our code?

By which I mean, should we teach `expand` to handle this, or is it not correct to be able to do this in the first place?

**Comment by mpoffald** [comment]
I am experimenting locally with an `expand` that will look for either `@type` or `type`, just to see if that fixes this issue, and it doesn't.

It's a small change, and if it _should_ be legal to use either, then maybe we should keep it even if it doesn't fix this issue.

**Comment by bplatz** [comment]
Its a good question... our use of :type-key is internal and is an optimization so we don't need do multiple `get` operations while processing. Understanding a node's type might affect the context expansion, so checking if the `@type` is going to affect the expansion context needs to be done before processing the node.

The best place to see what works and *should* be standard is always https://json-ld.org/playground/ - this is supposed to be the most spec-compliant implementation but that doesn't mean there are zero bugs... just can give a quick data point however if something works or doesn't work there.

**Comment by aaj3f** [comment]
@mpoffald @cap10morgan even removing the mapping between `type` and `@type` and `id` and `@id`, the issue still seems to exist that the properties I describe above `sh:path`, `sh:targetClass`, and `sh:datatype` aren't being resolved as references to other nodes, and so don't get evaluated properly in order to enforce the SHACL shape. I should clarify that, if I make that explicit in the transaction (like the following), then the SHACL shape enforcement works correctly, but the SHACL test in DB that I mention in the code snippet above doesn't require that same explicit syntax to function correctly:
```json
{
    "ledger": "shacl/data-type",
    "context": {
        "xsd": "http://www.w3.org/2001/XMLSchema#",
        "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
        "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
        "sh": "http://www.w3.org/ns/shacl#",
        "schema": "http://schema.org/",
        "skos": "http://www.w3.org/2008/05/skos#",
        "wiki": "https://www.wikidata.org/wiki/",
        "f": "https://ns.flur.ee/ledger#",
        "ex": "http://example.org/ns/"
    },
    "txn": {
        "@id": "ex:UserShape",
        "@type": ["sh:NodeShape"],
        "sh:targetClass": {"@id": "ex:User"},
        "sh:property": [
            {
                "sh:path": {"@id": "schema:name"},
                "sh:datatype": {"@id": "xsd:string"}
            }
        ]
    }
}
```

**Comment by mpoffald** [comment]
>  sh:path, sh:targetClass, and sh:datatype aren't being resolved as references to other nodes, and so don't get evaluated properly in order to enforce the SHACL shape. I

@aaj3f Yes, that's what I've found and have been investigating. Our handling of node values that are keywords vs strings is very different, leading to expansion results.

**Comment by mpoffald** [comment]
Specifically, this condition is present when parsing node values that are strings, but not for keywords: https://github.com/fluree/json-ld/blob/main/src/fluree/json_ld/impl/expand.cljc#L137

**Comment by mpoffald** [comment]
To clarify: this makes sense, we _do_ need a way to differentiate between a string that's a value and one that's a ref, but the way we do this now isn't catching the latter in the case of the SHACL shape.

**Comment by mpoffald** [comment]
Ok I just took a look at the [json-ld spec](https://www.w3.org/TR/2014/REC-json-ld-20140116/#advanced-concepts) as it pertains to embedding nodes, and I think the way this behaves now is correct. 

Because keywords can't be values, we can always assume those values are actually iris, but strings are ambiguous, so it needs to be explicit.

We could choose to attempt some inferencing to guess whether something is a ref (which has a convenience vs correctness tradeoff), but otherwise I think we may just need to require that `@id`

**Comment by mpoffald** [comment]
@bplatz @aaj3f I'm not sure there's anything that needs to be done here, what do you think? Requiring that `@id` to make a ref to another node seems in line with the spec.

re: the `:type-key` discussion, AFAICT this was just a red herring. As @bplatz said, that's just an optimization; a node using the expanded `@type` despite having a mapping of `{"type" : "@type"}` in the context  (and therefore a `:type-key` of `"type"`) was not being handled incorrectly from what I could tell. We just weren't leveraging that optimization in that case, so we did some extra work.

Using the expanded `@type` form in the node despite it being mapped in the context _is_ valid, _and_ it already worked fine in our json-ld lib. (we even [had an existing test for it](https://github.com/fluree/json-ld/blob/main/test/fluree/json_ld/impl/expand_test.cljc#L50), which I missed initially).

I do have a branch in the `json-ld` lib that extends the `:type-key` optimization to look for the expanded form of the `type-key` too, because I tested that change during my investigation of this issue. I could make a PR for it (since I've got it handy), but it doesn't have anything to do with this issue/SHACL specifically.

**Comment by mpoffald** [comment]
My take on this: this ticket originally arose as a high-priority item because we thought SHACL shapes were just not working, and had that been the case, it would've been a blocker. However, we _did_ have a correct way to get SHACL shape enforcement all along (https://github.com/fluree/http-api-gateway/issues/36#issuecomment-1458900496), we were just confused about the implications of using strings vs. keywords.

My vote, then, would be to close this issue (there was no bug, just a misunderstanding), and if we want to consider some inference work for convenience, create a new ticket for that and prioritize it accordingly.

**Comment by cap10morgan** [comment]
> My vote, then, would be to close this issue (there was no bug, just a misunderstanding), and if we want to consider some inference work for convenience, create a new ticket for that and prioritize it accordingly.

That sounds right to me. I say close.

**Comment by aaj3f** [comment]
This all sounds right to me @mpoffald & @cap10morgan
